Description: Use SDL semaphores instead of named semaphores, that fail under strict confinement
Forwarded: no
Author: Olivier Tilloy <olivier@tilloy.net>
Last-Update: 2017-05-15

--- a/source/tools/atlas/GameInterface/MessagePasserImpl.h
+++ b/source/tools/atlas/GameInterface/MessagePasserImpl.h
@@ -20,6 +20,7 @@
 
 #include "MessagePasser.h"
 
+#include "lib/external_libraries/libsdl.h"
 #include "ps/ThreadUtil.h"
 #include "ps/CStr.h"
 #include <queue>
@@ -40,8 +41,7 @@ public:
 
 private:
 	CMutex m_Mutex;
-	CStr m_SemaphoreName;
-	sem_t* m_Semaphore;
+	SDL_sem* m_Semaphore;
 	std::queue<AtlasMessage::IMessage*> m_Queue;
 	bool m_Trace;
 };

--- a/source/tools/atlas/GameInterface/MessagePasserImpl.cpp
+++ b/source/tools/atlas/GameInterface/MessagePasserImpl.cpp
@@ -32,39 +32,10 @@ using namespace AtlasMessage;
 MessagePasserImpl::MessagePasserImpl()
 : m_Trace(false), m_Semaphore(NULL)
 {
-	int tries = 0;
-	while (tries++ < 16) // some arbitrary cut-off point to avoid infinite loops
-	{
-		static char name[64];
-		sprintf_s(name, ARRAY_SIZE(name), "/wfg-atlas-msgpass-%d-%d",
-				(int)rand(1, 1000), (int)(time(0)%1000));
-		sem_t* sem = sem_open(name, O_CREAT | O_EXCL, 0700, 0);
-
-		// This cast should not be necessary, but apparently SEM_FAILED is not
-		// a value of a pointer type
-		if (sem == (sem_t*)SEM_FAILED || !sem)
-		{
-			int err = errno;
-			if (err == EEXIST)
-			{
-				// Semaphore already exists - try another one
-				continue;
-			}
-			// Otherwise, it's a probably-fatal error
-			debug_warn(L"sem_open failed");
-			break;
-		}
-		// Succeeded - use this semaphore
-		m_Semaphore = sem;
-		m_SemaphoreName = name;
-		break;
-	}
-
-	if (! m_Semaphore)
+	m_Semaphore = SDL_CreateSemaphore(0);
+	if (m_Semaphore == NULL)
 	{
 		debug_warn(L"Failed to create semaphore for Atlas - giving up");
-		// We will probably crash later - maybe we could fall back on sem_init, if this
-		// ever fails in practice
 	}
 }
 
@@ -72,9 +43,7 @@ MessagePasserImpl::~MessagePasserImpl()
 {
 	if (m_Semaphore)
 	{
-		// Clean up
-		sem_close(m_Semaphore);
-		sem_unlink(m_SemaphoreName.c_str());
+		SDL_DestroySemaphore(m_Semaphore);
 	}
 }
 
@@ -165,19 +134,7 @@ void MessagePasserImpl::Query(QueryMessa
 // 	while (0 != (err = sem_timedwait(psem, &abs_timeout)))
 // #endif
 
-	while (0 != sem_wait(m_Semaphore))
-	{
-		// If timed out, call callback and try again
-// 		if (errno == ETIMEDOUT)
-// 			timeoutCallback();
-// 		else
-		// Keep retrying while EINTR, but other errors are probably fatal
-		if (errno != EINTR)
-		{
-			debug_warn(L"Semaphore wait failed");
-			return; // (leaks the semaphore)
-		}
-	}
+	SDL_SemWait(m_Semaphore);
 
 	// Clean up
 	qry->m_Semaphore = NULL;

--- a/source/tools/atlas/GameInterface/GameLoop.cpp
+++ b/source/tools/atlas/GameInterface/GameLoop.cpp
@@ -184,7 +184,7 @@ static void* RunEngine(void* data)
 				{
 					// For queries, we need to notify MessagePasserImpl::Query
 					// that the query has now been processed.
-					sem_post((sem_t*) static_cast<QueryMessage*>(msg)->m_Semaphore);
+					SDL_SemPost((SDL_sem*) static_cast<QueryMessage*>(msg)->m_Semaphore);
 					// (msg may have been destructed at this point, so don't use it again)
 
 					// It's quite possible that the querier is going to do a tiny
